File: app/src/main/java/com/github/Doomsdayrs/apps/shosetsu/settings/SettingsController.java
package com.github.Doomsdayrs.apps.shosetsu.settings;

import android.content.SharedPreferences;
import android.graphics.Color;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;

import com.github.Doomsdayrs.apps.shosetsu.database.Database;

import org.json.JSONObject;

/**
 * This file is part of Shosetsu.
 * Shosetsu is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * Foobar is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * You should have received a copy of the GNU General Public License
 * along with Shosetsu.  If not, see https://www.gnu.org/licenses/ .
 * ====================================================================
 * Shosetsu
 * 14 / June / 2019
 *
 * @author github.com/doomsdayrs
 */
public class SettingsController {
    public static SharedPreferences download;
    public static SharedPreferences view;
    public static SharedPreferences advanced;
    public static SharedPreferences tracking;

    public static void init() {
        Settings.ReaderTextColor = view.getInt("ReaderTextColor", Color.BLACK);
        Settings.ReaderTextBackgroundColor = view.getInt("ReaderBackgroundColor", Color.WHITE);
    }

    public static boolean isOnline() {
        NetworkInfo activeNetwork = Settings.connectivityManager.getActiveNetworkInfo();
        if (activeNetwork != null)
            return activeNetwork.getType() == ConnectivityManager.TYPE_WIFI || activeNetwork.getType() == ConnectivityManager.TYPE_MOBILE;
        return false;
    }

    public static boolean isReaderLightMode() {
        return Settings.ReaderTextColor == Color.BLACK;
    }

    private static void setReaderColor(int text, int background) {
        Settings.ReaderTextColor = text;
        Settings.ReaderTextBackgroundColor = background;
        view.edit()
                .putInt("ReaderTextColor", text)
                .putInt("ReaderBackgroundColor", background)
                .apply();
    }

    public static void swapReaderColor() {
        if (isReaderLightMode())
            setReaderColor(Color.WHITE, Color.BLACK);
        else
            setReaderColor(Color.BLACK, Color.WHITE);
    }

    public static boolean isBookMarked(String chapterURL) {
        return Database.isBookMarked(chapterURL);
    }

    public static int getYBookmark(String chapterURL) {
        return Database.getBookmarkObject(chapterURL);
    }

    /**
     * Toggles bookmark
     *
     * @param chapterURL imageURL of chapter
     * @param saveData   JSON object containing scroll position and others
     * @return true means added, false means removed
     */
    public static boolean toggleBookmarkChapter(String chapterURL, JSONObject saveData) {
        if (isBookMarked(chapterURL)) {
            return !Database.removeBookMarked(chapterURL);
        } else {
            return Database.addBookMark(chapterURL, saveData);
        }
    }

    public static boolean isTrackingEnabled() {
        return tracking.getBoolean("enabled", false);
    }

    public static void addTracker() {
    }
}

File: app/src/main/java/com/github/doomsdayrs/apps/shosetsu/ui/settings/SettingsController.kt
package com.github.doomsdayrs.apps.shosetsu.ui.settings

import android.view.View
import com.github.doomsdayrs.apps.shosetsu.R
import com.github.doomsdayrs.apps.shosetsu.common.SettingsCard
import com.github.doomsdayrs.apps.shosetsu.ui.settings.SettingsController.Types.*
import com.github.doomsdayrs.apps.shosetsu.ui.settings.adapter.SettingsAdapter
import com.github.doomsdayrs.apps.shosetsu.view.base.RecyclerController

/*
 * This file is part of Shosetsu.
 *
 * Shosetsu is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Shosetsu is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Shosetsu.  If not, see <https://www.gnu.org/licenses/>.
 */

/**
 * Shosetsu
 * 9 / June / 2019
 */
class SettingsController : RecyclerController<SettingsAdapter, SettingsCard>() {
	override val layoutRes: Int = R.layout.settings
	override val viewTitleRes: Int = R.string.settings

	init {
		recyclerArray.clear()
		recyclerArray.addAll(arrayListOf(
				SettingsCard(DOWNLOAD),
				SettingsCard(UPDATE),
				SettingsCard(VIEW),
				SettingsCard(READER),
				SettingsCard(ADVANCED),
				SettingsCard(INFO),
				SettingsCard(BACKUP)
		))
	}

	override fun onViewCreated(view: View) {
	}

	override fun difAreItemsTheSame(oldItem: SettingsCard, newItem: SettingsCard): Boolean =
			oldItem.id == newItem.id

	override fun setupRecyclerView() {
		super.setupRecyclerView()
		recyclerView?.setHasFixedSize(true)
	}

	override fun createRecyclerAdapter(): SettingsAdapter =
			SettingsAdapter(recyclerArray, router)

	enum class Types {
		DOWNLOAD, // Settings for download options
		UPDATE,
		VIEW, // Settings for application appearance
		ADVANCED, // Settings that control more advanced application features
		INFO,  // Information of the app
		BACKUP, // Settings for backup and restoring data
		READER // Settings for reading novels in application
	}
}